src/workers/python/init.py.ts

```diff
@@ -1,44 +1,118 @@
+    def __init__(self, cb):
+        self.cb = cb
+        self.line = ""
+        self.override_builtins()
+
+    def message(self, data):
+        return self.cb(to_js(data))
+
+    def override_builtins(self):
+        self.override_output()
+        self.override_input()
 
-def __capture_stdout(cb):
-    class _OutputWriter:
-        def __init__(self):
-            self.encoding = "utf-8"
-            
-        def write(self, s):
-            cb(to_js({"type": "output", "data":s}))
-
-        def flush(self):
-            pass # Given data is always immediately written
-
-    sys.stdout = _OutputWriter()
-
-__papyros_input = ""
-def __override_stdin(cb):
-
-    def __papyros_input(prompt=""):
-        return __papyros_readline(prompt=prompt)[:-1] # Remove newline
-
-    def __papyros_readline(n=-1, prompt=""):
-        global __papyros_input
-        if not __papyros_input:
-            __papyros_input = cb(to_js({"type": "input", "data": prompt})) + "\\n"
-        if n < 0 or n > len(__papyros_input):
-            n = len(__papyros_input)
-        to_return = __papyros_input[0:n]
-        __papyros_input = __papyros_input[n:]
+    def override_output(self):
+        class __OutputWriter:
+            def __init__(self, type, on_write):
+                self.encoding = "utf-8"
+                self.type = type
+                self.on_write = on_write
+                
+            def write(self, s):
+                self.on_write(dict(type=self.type, data=s))
+
+            def flush(self):
+                pass # Given data is always immediately written
+        on_write = lambda d: self.message(d)
+        sys.stdout = __OutputWriter("output", on_write)
+        sys.stderr = __OutputWriter("error", on_write)
+
+    def readline(self, n=-1, prompt=""):
+        if not self.line:
+            self.line = self.message(dict(type="input", data=prompt)) + "\\n"
+        if n < 0 or n > len(self.line):
+            n = len(self.line)
+        to_return = self.line[0:n]
+        self.line = self.line[n:]
         return to_return
 
-    global input
-    input = __papyros_input
-    sys.stdin.readline = __papyros_readline
+    def override_input(self):
+        global input
+        input = lambda prompt="": self.readline(prompt=prompt)[:-1] # Remove newline
+
+        sys.stdin.readline = self.readline
+
+def clean_traceback(tb, filename):
+    # For some reason the first character of __file__ is lost in friendly_traceback
+    for prefix in ["\\"", "\\'", " "]:
+        tb = tb.replace(prefix + filename[1:], prefix + filename)
+    return html.escape(tb)
+
+def format_exception(filename, exc):
+    fr = FriendlyTraceback(type(exc), exc, exc.__traceback__)
+    fr.assign_generic()
+    fr.assign_cause()
+    tb = fr.info.get("simulated_python_traceback", "No traceback")
+    info = fr.info.get("generic", "No information available.")
+    why = fr.info.get("cause", "Unknown cause")
+    suggestions = fr.info.get("suggest", "No suggestions")
```

Comment:
```suggestion
    suggestions = fr.info.get("suggest", "")
```

The dangling "No suggestions" looks weird to me. I imagine similar treatment for the other data might be good.

![Screenshot from 2021-12-17 13-01-15](https://user-images.githubusercontent.com/3627481/146535737-e8a9572a-d4c4-458b-8c96-0dab9240f98a.png)

Reply:
I added them temporarily as one of the top questions in this PR is what fallback to use. I'm waiting for my promotor to take a look and provide suggestions.